src/crypto/PrivateKeyBundle.ts

```diff
@@ -98,159 +100,142 @@ export default class PrivateKeyBundle implements proto.PrivateKeyBundleV1 {
-  try {
-    const b = proto.EncryptedPrivateKeyBundle.decode(bytes)
-    return [b.v1, false]
-  } catch (e) {
-    if (
-      e instanceof RangeError ||
-      (e instanceof Error && e.message.startsWith('invalid wire type'))
-    ) {
-      // Adds a default fallback for older versions of the KeyBundles
-      return [proto.EncryptedPrivateKeyBundleV1.decode(bytes), true]
+  // sharedSecret derives a secret from peer's key bundles using a variation of X3DH protocol
+  // where the sender's ephemeral key pair is replaced by the sender's pre-key.
+  // @peer is the peer's public key bundle
+  // @myPreKey indicates which of my preKeys should be used to derive the secret
+  // @recipient indicates if this is the sending or receiving side.
+  async sharedSecret(
+    peer: PublicKeyBundle,
+    myPreKey: PublicKey,
+    isRecipient: boolean
+  ): Promise<Uint8Array> {
+    if (!peer.identityKey || !peer.preKey) {
+      throw new Error('invalid peer key bundle')
+    }
+    if (!(await peer.identityKey.verifyKey(peer.preKey))) {
+      throw new Error('peer preKey signature invalid')
+    }
+    if (!this.identityKey) {
+      throw new Error('missing identity key')
+    }
+    let dh1: Uint8Array, dh2: Uint8Array, preKey: PrivateKey
+    if (isRecipient) {
+      preKey = this.findPreKey(myPreKey)
+      dh1 = preKey.sharedSecret(peer.identityKey)
+      dh2 = this.identityKey.sharedSecret(peer.preKey)
+    } else {
+      preKey = this.findPreKey(myPreKey)
+      dh1 = this.identityKey.sharedSecret(peer.preKey)
+      dh2 = preKey.sharedSecret(peer.identityKey)
     }
-    throw new Error("Couldn't decode encrypted bundle:" + e)
+    const dh3 = preKey.sharedSecret(peer.preKey)
+    const secret = new Uint8Array(dh1.length + dh2.length + dh3.length)
+    secret.set(dh1, 0)
+    secret.set(dh2, dh1.length)
+    secret.set(dh3, dh1.length + dh2.length)
+    return secret
+  }
+
+  encode(): Uint8Array {
+    return proto.PrivateKeyBundle.encode({ v1: this, v2: undefined }).finish()
   }
 }
 
-// getPrivateV1Bundle returns the decoded bundle from the provided bytes. If there is an error decoding the bundle it attempts
-// to decode the bundle as a legacy bundle. Additionally return whether the bundle is in the expected format.
-function getPrivateV1Bundle(
-  bytes: Uint8Array
-): [proto.PrivateKeyBundleV1 | undefined, boolean] {
-  try {
-    const b = proto.PrivateKeyBundle.decode(bytes)
-    return [b.v1, false]
-  } catch (e) {
-    if (
-      e instanceof RangeError ||
-      (e instanceof Error && e.message.startsWith('invalid wire type'))
-    ) {
-      // Adds a default fallback for older versions of the proto
-      return [proto.PrivateKeyBundleV1.decode(bytes), true]
-    }
-    throw new Error("Couldn't decode private bundle:" + e)
+export type PrivateKeyBundle = PrivateKeyBundleV1 | PrivateKeyBundleV2
+
+export function DecodePrivateKeyBundle(bytes: Uint8Array): PrivateKeyBundle {
```

Comment:
One thing I only noticed after review is the casing of `DecodeContactBundle`. Conventionally, Javascript functions are camelCased and not PascalCased (with Pascal Case reserved for classes and constructors). Think we should stick with that convention

Reply:
Thanks for calling this out, I have no clue about the conventions, so please do so whenever you see something like this.